### PR TITLE
feat: lenient apparmor profiles

### PR DESCRIPTION
**Pull request by PrimalPimmy**
**Purpose of PR?**:

This PR is part of toning down apparmor restrictions on Kubearmor. Then we can use that to secure Kubearmor by applying KSPs as a privileged container.
 
**Does this PR introduce a breaking change?**
Could be, it will be privileged after all.

**If the changes in this PR are manually verified, list down the scenarios covered:**:

**Additional information for reviewer?** :
_Mention if this PR is part of any design or a continuation of previous PRs_

**Checklist:**
- [ ] Bug fix. Fixes #<issue number>
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [x] Commit has unit tests
- [x] Commit has integration tests

### CODE REVIEW

File: KubeArmor/core/containerdHandler.go [hunk 1]
```diff
@@ -7,11 +7,14 @@ package core
 import (
 	"context"
 	"fmt"
+	"github.com/containerd/typeurl/v2"
 	"os"
 	"strconv"
 	"strings"
 	"time"
 
+	"golang.org/x/exp/slices"
```
Comment (daemon1024) [inline KubeArmor/core/containerdHandler.go:16]: slices package is stable now, we don't need the exp import now. https://pkg.go.dev/slices

Comment (DelusionalOptimist) [reply KubeArmor/core/containerdHandler.go:16]: It moved out of experimental in go1.21. We'll need to update go version.

Comment (DelusionalOptimist) [reply KubeArmor/core/containerdHandler.go:16]: IMO we should do it ASAP. Dev environments which use go 1.21 don't work well with projects using go1.20.
Running things like go mod tidy adds weird directives which don't work with older go versions. Even gopls doesn't offer any completion/documentation for older projects.

Comment (PrimalPimmy) [reply KubeArmor/core/containerdHandler.go:16]: I think we should have a new PR for updating go versions.

Comment (daemon1024) [reply KubeArmor/core/containerdHandler.go:16]: Can you open an issue, we can resolve the comment then

File: KubeArmor/core/unorchestratedUpdates.go [hunk 2]
```diff
@@ -524,12 +531,20 @@ func (dm *KubeArmorDaemon) ParseAndUpdateContainerSecurityPolicy(event tp.K8sKub
...
 			for idx, ctr := range dm.Containers {
 				if ctr.ContainerName == containername {
 					newPoint.Containers = append(newPoint.Containers, ctr.ContainerID)
 					ctr.NamespaceName = newPoint.NamespaceName
 					ctr.EndPointName = newPoint.EndPointName
+					/*
+						if ctr.Privileged {
+							newPoint.PrivilegedContainers[containername] = struct{}{}
+						}
+					*/
```
Comment (daemon1024) [inline KubeArmor/core/unorchestratedUpdates.go]:
```suggestion
```

File: KubeArmor/core/unorchestratedUpdates.go [hunk 3]
```diff
@@ -430,9 +430,16 @@ func (dm *KubeArmorDaemon) ParseAndUpdateContainerSecurityPolicy(event tp.K8sKub
...
+	var privilegedProfiles map[string]struct{}
 	for idx, endPoint := range dm.EndPoints {
 		endPointIndex++
 
+		/*
+			if _, ok := endPoint.PrivilegedContainers[containername]; ok {
+				privilegedProfiles[appArmorAnnotations[containername]] = struct{}{}
+			}
+		*/
+
```
Comment (daemon1024) [inline KubeArmor/core/unorchestratedUpdates.go]:
```suggestion

```

File: KubeArmor/enforcer/appArmorEnforcer.go [hunk 4]
```diff
@@ -454,7 +479,16 @@ func (ae *AppArmorEnforcer) UnregisterAppArmorHostProfile() bool {
...
 func (ae *AppArmorEnforcer) UpdateAppArmorProfile(endPoint tp.EndPoint, appArmorProfile string, securityPolicies []tp.SecurityPolicy) {
-	if policyCount, newProfile, ok := ae.GenerateAppArmorProfile(appArmorProfile, securityPolicies, endPoint.DefaultPosture); ok {
+
+	ae.AppArmorPrivilegedProfilesLock.Lock()
+	_, privileged := ae.AppArmorPrivilegedProfiles[appArmorProfile]
+	ae.AppArmorPrivilegedProfilesLock.Unlock()
+
+	if privileged {
+		kg.Warnf("IS PRIVILEGED CONTAINER PROFILE apparmorProfile: %s", appArmorProfile)
+	}
```
Comment (daemon1024) [inline KubeArmor/enforcer/appArmorEnforcer.go]: We should not Warn here, let's remove this or use Debug if we need it

File: KubeArmor/enforcer/appArmorEnforcer.go [hunk 5]
```diff
@@ -508,6 +542,7 @@ func (ae *AppArmorEnforcer) UpdateSecurityPolicies(endPoint tp.EndPoint) {
 
 	for _, appArmorProfile := range endPoint.AppArmorProfiles {
 		if kl.ContainsElement([]string{"docker-default", "unconfined", "cri-containerd.apparmor.d", "crio-default", ""}, appArmorProfile) {
+			kg.Warnf("CONTAINER SKIPPED apparmorprofile %s", appArmorProfile)
```
Comment (daemon1024) [inline KubeArmor/enforcer/appArmorEnforcer.go]: Remove/Debug Log

File: KubeArmor/enforcer/appArmorEnforcer.go [hunk 6]
```diff
@@ -454,7 +479,16 @@ func (ae *AppArmorEnforcer) UnregisterAppArmorHostProfile() bool {
 
 // UpdateAppArmorProfile Function
 func (ae *AppArmorEnforcer) UpdateAppArmorProfile(endPoint tp.EndPoint, appArmorProfile string, securityPolicies []tp.SecurityPolicy) {
-	if policyCount, newProfile, ok := ae.GenerateAppArmorProfile(appArmorProfile, securityPolicies, endPoint.DefaultPosture); ok {
+
+	ae.AppArmorPrivilegedProfilesLock.Lock()
+	_, privileged := ae.AppArmorPrivilegedProfiles[appArmorProfile]
+	ae.AppArmorPrivilegedProfilesLock.Unlock()
```
Comment (daemon1024) [inline KubeArmor/enforcer/appArmorEnforcer.go:487]: Let's add a comment here explaining what's the reason for checking privileged and what do we do differently

File: deployments/go.sum [hunk 7]
```diff
@@ -87,20 +87,73 @@ golang.org/x/net v0.0.0-20190404232315-eb5bcb51f2a3/go.mod h1:t9HGtf8HONx5eT2rtn
 golang.org/x/net v0.0.0-20190620200207-3b0461eec859/go.mod h1:z5CRVTTTmAJ677TzLLGU+0bjPO0LkuOLi4/5GtJWs/s=
 golang.org/x/net v0.0.0-20200226121028-0de0cce0169b/go.mod h1:z5CRVTTTmAJ677TzLLGU+0bjPO0LkuOLi4/5GtJWs/s=
 golang.org/x/net v0.0.0-20201021035429-f5854403a974/go.mod h1:sp8m0HH+o8qH0wwXwYZr8TS3Oi6o0r6Gce1SSxlDquU=
+<<<<<<< HEAD
```
Comment (daemon1024) [inline deployments/go.sum]: Failed git merge, please fix

File: getting-started/kubearmor_apparmor_implementation_overview.md [hunk 8]
```diff
@@ -24,4 +24,13 @@ The domain transion is made possible using the `px` permission in apparmor.
 
 ![Action Conflict](./resources/Apparmor-domain-transition.png)
 
-The above figure illustrate how we decide that we need to perform a domain transition for a certain binary.
\ No newline at end of file
+The above figure illustrate how we decide that we need to perform a domain transition for a certain binary.
+
+## Apparmor profiles for privileged pods and pods with additional capabilities 
+
+Pods with additional capabilities are supported across all runtimes, hence apparmor profiles can be set here.
```
Comment (daemon1024) [inline getting-started/kubearmor_apparmor_implementation_overview.md]: Let's document, we provide more capabilities than docker default profile to make them work

**Review by daemon1024 — CHANGES_REQUESTED** [review]
Please find comments inline

File: contribution/k3s/install_k3s.sh [hunk 9]
```diff
@@ -16,10 +16,10 @@ fi
 if [ "$RUNTIME" == "docker" ]; then # docker
     CGROUP_SYSTEMD=$(docker info 2> /dev/null | grep -i cgroup | grep systemd | wc -l)
     if [ $CGROUP_SYSTEMD == 1 ]; then
-        curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION="v1.23.9+k3s1" K3S_KUBECONFIG_MODE="644" INSTALL_K3S_EXEC="--disable=traefik --docker --kubelet-arg cgroup-driver=systemd" sh -
+        curl -sfL https://get.k3s.io | K3S_KUBECONFIG_MODE="644" INSTALL_K3S_EXEC="--disable=traefik --docker --container-runtime-endpoint unix:///var/run/docker.sock --kubelet-arg cgroup-driver=systemd" sh -
```
Comment (daemon1024) [inline contribution/k3s/install_k3s.sh:19]: Might be due to version upgrade now, the apparmor profile registration might be slower? Out of control for us but might be the reason for recent rise in failure specific to your PR

**Review by daemon1024 — APPROVED** [review]

**Review by DelusionalOptimist — APPROVED** [review]

**Pull request merged**